src/store/singing.ts

```diff
@@ -251,6 +258,96 @@ export const singingStore = createPartialStore<SingingStoreTypes>({
     },
   },
 
+  ADD_NOTE: {
+    mutation(state, { note }: { note: Note }) {
+      if (state.score) {
+        const notes = state.score.notes.concat(note).sort((a, b) => {
+          return a.position - b.position;
+        });
+        state.score.notes = notes;
+      }
+    },
+    // ノートを追加する
+    // NOTE: 重複削除など別途追加
+    async action({ state, commit }, { note }: { note: Note }) {
+      if (state.score === undefined) {
+        throw new Error("Score is not initialized.");
+      }
+      commit("ADD_NOTE", { note });
+    },
+  },
+
+  CHANGE_NOTE: {
+    mutation(state, { index, note }: { index: number; note: Note }) {
+      state.score?.notes.splice(index, 1, note);
+    },
+    async action(
+      { state, commit },
+      { index, note }: { index: number; note: Note }
+    ) {
+      if (state.score === undefined) {
+        throw new Error("Score is not initialized.");
+      }
+      commit("CHANGE_NOTE", { index, note });
+    },
+  },
+
+  REMOVE_NOTE: {
+    mutation(state, { index }: { index: number }) {
+      state.score?.notes.splice(index, 1);
```

Comment:
パフォーマンスを計測して気づいたのですが、配列のstateに直接spliceを使用するとreactiveEffectが大量に発生してしまうみたいです。。一度配列をコピーして、操作を行ってからstateにセットするようにした方が良いかもしれません。（テンポのmutationも後で修正したいと思います…！）
```suggestion
      if (state.score) {
        const notes = [...state.score.notes];
        notes.splice(index, 1);
        state.score.notes = notes;
      }
```

Reply:
@sigprogramming 
ありがとうございます！
そうでなくともstate内の値は直接操作まずかった記憶もある(VuexではなくReduxにおけるものだったかも…)ので、あわせて修正いたします！

ほか、reactiveEffectとは若干異なりますが、

- グリッドとかは`v-memo`でズーム時以外memo化すればマシになるかも → Vue3.2以降サポートなので無理そう？
- IntersectionObserverで見えてる部分だけ描画
- virtual-scrollの導入

などパフォーマンス最適化はいろいろできそうかもですが、
のちほどの課題にできればと考えています…！

Reply:
修正ありがとうございます！
パフォーマンス最適化について、承知しました！